src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -0,0 +1,513 @@
+                };
+                edits.Add(edit);
+            }
+            else if (firstNonWhitespaceOffset != desiredIndentation.Length)
+            {
+                // This means the '}' is on its own line but is not indented correctly. Correct it.
+                var edit = new TextEdit()
+                {
+                    Range = new Range(
+                    new Position(closeCurlyLine.LineNumber, 0),
+                    new Position(closeCurlyLine.LineNumber, firstNonWhitespaceOffset)),
+                    NewText = desiredIndentation
+                };
+                edits.Add(edit);
+            }
+        }
+
+        private SourceText ApplyChanges(SourceText original, IEnumerable<TextChange> changes)
+        {
+            var changed = original.WithChanges(changes);
+            return changed;
+        }
+
+        private void TrackChangeInSpan(SourceText oldText, TextSpan originalSpan, SourceText newText, out TextSpan changedSpan, out TextSpan changeEncompassingSpan)
+        {
+            var affectedRange = newText.GetEncompassingTextChangeRange(oldText);
+            changeEncompassingSpan = affectedRange.Span;
+            if (!originalSpan.Contains(changeEncompassingSpan))
+            {
+                _logger.LogDebug($"The changed region {changeEncompassingSpan} was not a subset of the span {originalSpan} being tracked. This is unexpected.");
+            }
+
+            changedSpan = TextSpan.FromBounds(originalSpan.Start, originalSpan.End + affectedRange.NewLength - affectedRange.Span.Length);
+        }
+
+        private TextChange[] Diff(SourceText oldText, SourceText newText, TextSpan? spanToDiff = default)
+        {
+            // Once https://github.com/dotnet/roslyn/issues/41413 is fixed,
+            // the following lines can be replaced with `newText.GetTextChanges(oldText)`.
+
+            var spanToTrack = spanToDiff ?? TextSpan.FromBounds(0, oldText.Length);
+            TrackChangeInSpan(oldText, spanToTrack, newText, out var changedSpanToTrack, out _);
+            var change = new TextChange(spanToTrack, newText.GetSubText(changedSpanToTrack).ToString());
+            return new[] { change };
+        }
+
+        private static TextSpan GetSpanIncludingPrecedingWhitespaceInLine(SourceText sourceText, int start, int end)
+        {
+            var line = sourceText.Lines.GetLineFromPosition(start);
+            var precedingLineText = sourceText.GetSubTextString(TextSpan.FromBounds(line.Start, start));
+            var precedingWhitespaceLength = precedingLineText.GetTrailingWhitespace().Length;
+
+            return TextSpan.FromBounds(start - precedingWhitespaceLength, end);
+        }
+
+        private static string GetIndentationString(FormattingContext context, int indentationLevel)
+        {
+            var indentChar = context.Options.InsertSpaces ? ' ' : '\t';
+            var indentation = new string(indentChar, (int)context.Options.TabSize * indentationLevel);
```

Comment:
In the case of `InsertSpaces == false` wont this insert a bazillion tabs?

Reply:
It's upto the person to also reduce the TabSize accordingly.

Reply:
VSCode documents it that way?

Reply:
Okay I was totally wrong. I've had the wrong understanding of what tab size meant. I'll update this logic and add tests for this.